fix(ngcc): capture entry-point dependencies from typings as well as source

petebacondarwin opened the pull request:
See FW-1781 for the background

Fixes #34411

JoostK commented on packages/compiler-cli/ngcc/test/dependencies/dependency_resolver_spec.ts:
```diff
@@ -93,6 +99,12 @@ runInEachFileSystem(() => {
           [_('/third/index.js')]: {resolved: [fourth.path, _('/ignored-2')], missing: []},
           [_('/fourth/sub2/index.js')]: {resolved: [fifth.path], missing: []},
           [_('/fifth/index.js')]: {resolved: [], missing: []},
+          [_('/first/index.d.ts')]:
```
Would it be an idea to group the .js and .d.ts adjacent to each other? I would expect that to be more clear.

petebacondarwin replied:
I moved these into their own dependency arrays now that I have separated out the dts host from the src host in the test.

JoostK commented on packages/compiler-cli/ngcc/src/main.ts:
```diff
@@ -164,12 +164,16 @@ export function mainNgcc(
-      umd: umdDependencyHost,
-      commonjs: commonJsDependencyHost
-    });
+    const dtsDependencyHost = new EsmDependencyHost(
```
Also, I'm surprised the `EsmDependencyHost` works as is? Wouldn't the following line have to be changed to use `ts.ScriptKind.TS`, or is TypeScript lenient enough to parse .d.ts just fine when using `ts.ScriptKind.JS`?

https://github.com/angular/angular/blob/ae142a682783043890b3d795da0791d7070bc227/packages/compiler-cli/ngcc/src/dependencies/esm_dependency_host.ts#L41

petebacondarwin replied:
Yeah, just parsing the AST is fine  - `import` statements are valid JS.

JoostK commented:
Could you update the commit message to include the fixes #34411 reference?

petebacondarwin commented on packages/compiler-cli/ngcc/src/dependencies/dependency_host.ts:
```diff
@@ -11,7 +11,8 @@ import {resolveFileWithPostfixes} from '../utils';
 
 export interface DependencyHost {
-  findDependencies(entryPointPath: AbsoluteFsPath): DependencyInfo;
+  findDependencies(
```
Good idea!

petebacondarwin commented on packages/compiler-cli/ngcc/test/dependencies/dependency_resolver_spec.ts:
```diff
@@ -35,7 +35,7 @@ runInEachFileSystem(() => {
       moduleResolver = new ModuleResolver(fs);
       host = new EsmDependencyHost(fs, moduleResolver);
-      resolver = new DependencyResolver(fs, new MockLogger(), {esm5: host, esm2015: host});
+      resolver = new DependencyResolver(fs, new MockLogger(), {esm5: host, esm2015: host}, host);
```
It doesn't really matter in these tests but since it is a bit confusing, I will change it.

petebacondarwin committed:
fixup! refactor(ngcc): pass dependency info to `collectDependencies()`

petebacondarwin committed:
refactor(ngcc): pass dependency info to `collectDependencies()`

Rather than return a new object of dependency info from calls to
`collectDependencies()` we now pass in an object that will be updated
with the dependency info. This is in preparation of a change where
we will collect dependency information from more than one
`DependencyHost`.

Also to better fit with this approach the name is changed from
`findDependencies()` to `collectDependencies()`.

petebacondarwin committed:
refactor(ngcc): resolve modules based on the provided `moduleResolver`

The `DependencyHost` implementations were duplicating the "postfix" strings
which are used to find matching paths when resolving module specifiers.
Now the hosts reuse the postfixes given to the `ModuleResolver` that is
passed to the host.

petebacondarwin committed:
fixup! refactor(ngcc): pass dependency info to `collectDependencies()`

petebacondarwin committed:
fixup! refactor(ngcc): resolve modules based on the provided `moduleResolver`

petebacondarwin committed:
fix(ngcc): capture entry-point dependencies from typings as well as source

ngcc computes a dependency graph of entry-points to ensure that
entry-points are processed in the correct order. Previously only the imports
in source files were analysed to determine the dependencies for each
entry-point.

This is not sufficient when an entry-point has a "type-only" dependency
 - for example only importing an interface from another entry-point.
In this case the "type-only" import does not appear in the
source code. It only appears in the typings files. This can cause a
dependency to be missed on the entry-point.

This commit fixes this by additionally processing the imports in the
typings program, as well as the source program.

Note that these missing dependencies could cause unexpected flakes when
running ngcc in async mode on multiple processes due to the way that
ngcc caches files when they are first read from disk.

Fixes #34411

// FW-1781

petebacondarwin committed:
fixup! fix(ngcc): capture entry-point dependencies from typings as well as source

gkalpak reviewed (approved):
OOC, have you done any measurements regarding how much in the overall performance affected by this change?

petebacondarwin commented on packages/compiler-cli/ngcc/src/dependencies/dependency_host.ts:
```diff
@@ -11,7 +11,8 @@ import {resolveFileWithPostfixes} from '../utils';
 export interface DependencyHost {
-  findDependencies(entryPointPath: AbsoluteFsPath): DependencyInfo;
+  collectDependencies(
+      entryPointPath: AbsoluteFsPath, {dependencies, missing, deepImports}: DependencyInfo): void;
```
Is there any reason not to? 😛 
This is how it is implemented in concrete classes and so I guess I was just keeping them in sync.

alxhub replied:
Hmm, interesting. I actually had no idea this was legal.

petebacondarwin commented on packages/compiler-cli/ngcc/test/dependencies/dependency_resolver_spec.ts:
```diff
@@ -258,42 +324,78 @@ runInEachFileSystem(() => {
+        expect(dtsHost.collectDependencies)
+            .toHaveBeenCalledWith(fs.resolve(fourth.path, 'sub2/index.d.ts'), jasmine.any(Object));
+        expect(dtsHost.collectDependencies)
+            .toHaveBeenCalledWith(fs.resolve(fifth.path, 'index.d.ts'), jasmine.any(Object));
```
Does that add much value?
Given that we are using TypeScript, the compiler will ensure that the object has the correct layout.
And in this test I am only interested in the paths that are being passed to the function.

gkalpak replied:
Fair enough :smiley:

petebacondarwin commented:
RE performance change: note that this change only affects the dependency analysis phase, which is around 7.4% of the total processing time (using integration/cli-hello-world-ivy-compat). After this PR the average percentage goes up to around 7.8%.
